src/storage/mod.rs

```diff
@@ -1040,6 +1154,106 @@ impl Storage {
                         &Key::from_encoded(key),
                         limit,
                         &mut statistics,
+                        key_only,
+                    ).map_err(Error::from)
+                        .map(|r| {
+                            let mut valid_keys = 0;
+                            let mut bytes_read = 0;
+                            let mut stats = Statistics::default();
+                            r.iter().for_each(|r| {
+                                if let Ok(ref pair) = *r {
+                                    valid_keys += 1;
+                                    bytes_read += pair.0.len() + pair.1.len();
+                                }
+                            });
+                            stats.data.flow_stats.read_keys = valid_keys;
+                            stats.data.flow_stats.read_bytes = bytes_read;
+                            thread_ctx.collect_read_flow(ctx.get_region_id(), &stats);
+                            thread_ctx.collect_key_reads(CMD, valid_keys as u64);
+                            r
+                        });
+
+                    thread_ctx.collect_scan_count(CMD, &statistics);
+
+                    result
+                })
+                .then(move |r| {
+                    _timer.observe_duration();
+                    r
+                })
+        });
+
+        future::result(res)
+            .map_err(|_| Error::SchedTooBusy)
+            .flatten()
+    }
+
+    fn raw_batch_scan(
+        snapshot: Box<Snapshot>,
+        keys: Vec<Key>,
+        start_key_limit: usize,
+        total_limit: usize,
+        stats: &mut Statistics,
+        key_only: bool,
+    ) -> engine::Result<Vec<Result<KvPair>>> {
+        let mut cursor = snapshot.iter(IterOption::default(), ScanMode::Forward)?;
+        let mut pairs = vec![];
+        for key in &keys {
+            if !cursor.seek(key, &mut stats.data)? {
+                continue;
+            }
+            let mut key_pairs = 0;
+            while cursor.valid() && key_pairs < start_key_limit && pairs.len() < total_limit {
+                pairs.push(Ok((
+                    cursor.key().to_owned(),
+                    if key_only {
+                        vec![]
+                    } else {
+                        cursor.value().to_owned()
+                    },
+                )));
+                cursor.next(&mut stats.data);
+                key_pairs += 1;
+            }
+            if pairs.len() == total_limit {
+                break;
+            }
+        }
+        Ok(pairs)
+    }
+
+    pub fn async_raw_batch_scan(
```

Comment:
I think row_batch_scan is redundant @wentaoxu @breeswish ?

Reply:
raw_batch_scan is for scanning multiple not contiguous key ranges that fall into one region at once. It can replace single scan actually from my point of view.

Reply:
Is there any using case shows that the reducing RPC of batch scan can make some good profit?

Reply:
Sure. Here is the story. We were building a bigtable storage on top of TiKV. And the bigtable record is mapped into TiKV like this.
Key: column family id + row key + qualifier + timestamp
Value: value
Normally, user want to get data of several qualifiers of a row at once. But usually they don't care what version but just want to get the latest one. More concisely, they don't even know what version number is the latest.
To implement this kind of query, I want to scan records with several start key instead of either calling scan multiple times or scan too much unnecessary data and only keep whatever the data user has requested. And this is the reason behind. It sounds strange, I'm not sure if it's a common use cases other application might need.